### PR TITLE
Design matrix generation

### PR DESCRIPTION
**Pull request by TomMaullin**
This branch now generates the Design Matrix from scratch if using an FSL generated NIDM-Results pack. It also includes a test for checking the NIDM pack is not damaged when the mip is generated (issue #13).

### CODE REVIEW

File: test/nidmExampleDataTest.m [hunk 1]
```diff
@@ -59,5 +59,17 @@ function checkViewerRunsSPMwoMIP(testCase)
             nidm_results_display(fullfile(data_path, 'nidmwithoutMIP.json'));
         end
         
+        %Checking the nidm json is not damaged by the viewer.
+        function checkNIDMUnaffected(testCase)
+            nidm_results_display(fullfile(fileparts(mfilename('fullpath')), '..', 'Data', 'fsl_default', 'nidm.json'));
+            originalNIDM = spm_jsonread('C:\Users\owner\Documents\Project-NIDASH\nidmresults-spmhtml\Data\fsl_default\nidmWithoutMip.json');
```
Comment (cmaumet) [inline test/nidmExampleDataTest.m]: @TomMaullin: could you update this line (and the next one) to use a relative path? Currently this test can only run on your machine.

File: Data/fsl_default/nidm.json [hunk 2]
```diff
@@ -85,608 +85,623 @@
...
+        "x_value": "http://neurovault.org/collections/BNRPBNVV/fsl_default.nidm/ExcursionSet.nii.gz"
       },
-      "prov:wasGeneratedBy": {
-        "@id": "niiri:9833881f-a909-48dd-8930-a832048c9cfe"
+      "prov_wasGeneratedBy": {
+        "x_id": "niiri:9833881f-a909-48dd-8930-a832048c9cfe"
       },
-      "rdfs:label": "Excursion Set Map"
+      "rdfs_label": "Excursion Set Map",
+      "nidm_hasMaximumIntensityProjection": {
```
Comment (cmaumet) [inline Data/fsl_default/nidm.json]: The json file for fsl_default should not have a MIP. It looks like the original file was overwritten.

Comment (cmaumet) [inline Data/fsl_default/nidm.json]: Solved in 9997e0b.

**Comment by cmaumet** [comment]
Hi @TomMaullin,

I have just reviewed your PR and added a number of edits. See below for the details:
 - ece7863: Use relative paths in the `checkNIDMUnaffected` test rather than absolute paths.
 - 5494f32: Delete existing `index.html` before `checkNIDMUnaffected` is run (as for the other tests).
 - 9997e0b: Revert the json for the `fsl_default` example back to the original version (it looked like it had been overwritten by the one generated by SPM)
 - 89bc6a1: After writing the json back to the json file with `spm_jsonread`, replace all occurrence of `x_` by `@`. As Matlab does not handle well variable names starting with `@` those symbols are changed to `x_` by `spm_jsonread` but we should really change them back when writing to the file. This is something that ideally should be handled by `spm_jsonwrite`.

After those changes when I run the tests, I get the following error for `checkNIDMUnaffected`

```
>> tests = nidmExampleDataTest

tests = 

  nidmExampleDataTest with no properties.

>> run(tests, 'checkNIDMUnaffected')
Running nidmExampleDataTest

================================================================================
Verification failed in nidmExampleDataTest/checkNIDMUnaffected.

    ---------------------
    Framework Diagnostic:
    ---------------------
    verifyEqual failed.
    --> Path to failure:<Value>.x_type{1}
        --> The strings are not equal
        
        Actual String:
            nidm_NIDMResultsExport
        Expected String:
            prov:Activity
    
    Actual struct:
                              x_id: 'niiri:adc6bec6-f644-46b2-8736-1f134bba356a'
                            x_type: {2x1 cell}
            prov_wasAssociatedWith: [1x1 struct]
                        rdfs_label: 'NIDM-Results export'
    Expected struct:
                              x_id: 'niiri:adc6bec6-f644-46b2-8736-1f134bba356a'
                            x_type: {2x1 cell}
            prov_wasAssociatedWith: [1x1 struct]
                        rdfs_label: 'NIDM-Results export'

    ------------------
    Stack Information:
    ------------------
    In /Users/cmaumet/Projects/Data_sharing/dev/nidmresults-spmhtml/test/nidmExampleDataTest.m (nidmExampleDataTest.checkNIDMUnaffected) at 74
================================================================================
.
Done nidmExampleDataTest
__________

Failure Summary:

     Name                                     Failed  Incomplete  Reason(s)
    ======================================================================================
     nidmExampleDataTest/checkNIDMUnaffected    X                 Failed by verification.
    

ans = 

  TestResult with properties:

          Name: 'nidmExampleDataTest/checkNIDMUnaffected'
        Passed: 0
        Failed: 1
    Incomplete: 0
      Duration: 1.5784

Totals:
   0 Passed, 1 Failed, 0 Incomplete.
   1.5784 seconds testing time.

```

The test fails due to the ordering of the values for the `x_type` attributes. This should not matter in principle (the graphs are indeed equal) but the `verifyEqual` function does not know that order is irrelevant. Also, I have noticed that the json written out by `spm_jsonwrite` is not indented (all the text is on a single line) which is annoying. 

So, my suggestion is to comment out the update of the json using `spm_jsonwrite` until we have discussed those issues with Guillaume. This is done in c19e646.

After those changes, the tests pass:
```
>> runTest
Running nidmExampleDataTest
.....
Done nidmExampleDataTest
__________


result = 

  1x5 TestResult array with properties:

    Name
    Passed
    Failed
    Incomplete
    Duration

Totals:
   5 Passed, 0 Failed, 0 Incomplete.
   5.7703 seconds testing time.
```

**Comment by cmaumet** [comment]
@TomMaullin: can you review my changes and, if you are fine with them, merge this PR?

Starting up from a clean state (i.e. `git status` returns `nothing to commit, working directory clean`), to download my updates on your machine you can do:
```
git checkout designMatrixGeneration
git pull origin designMatrixGeneration
```
If the previous commands worked fine, when you do:
```
git log -6
```
you should get:
```
$ git log -6
commit c19e64653599b4617b9a6611b7628cbf5ea73a9c
Author: cmaumet <c.m.j.maumet@warwick.ac.uk>
Date:   Wed Nov 23 14:19:53 2016 +0000

    Commenting out update of the json file (to add MIP)

commit 89bc6a1ef75cfcceab9d3e6aaa896ac6a91f0102
Author: cmaumet <c.m.j.maumet@warwick.ac.uk>
Date:   Wed Nov 23 14:06:07 2016 +0000

    Replace x_ by @ in written json document

commit 9997e0b45b191ebd1886c7b323c2c1f29cfdf8c5
Author: cmaumet <c.m.j.maumet@warwick.ac.uk>
Date:   Wed Nov 23 12:26:51 2016 +0000

    Revert fsl_default json to original version

commit 5494f322f6b32bf2cba7779a898aeadd1d877d9c
Author: cmaumet <c.m.j.maumet@warwick.ac.uk>
Date:   Wed Nov 23 12:12:22 2016 +0000

    Delete index.html before running checkNIDMUnaffected test

commit ece7863bdf3cf3171a519e69a6a62c7a1fcf092f
Author: cmaumet <c.m.j.maumet@warwick.ac.uk>
Date:   Wed Nov 23 12:07:13 2016 +0000

    fix: use relative path to test data

commit fdb05bb07fe927b57dd46e449565ff1f17e416ed
Merge: 92cc270 108e428
Author: Tom Maullin <tommaullin@gmail.com>
Date:   Thu Nov 10 18:01:48 2016 +0000

    merge with mipGeneration
```

**Comment by gllmflndn** [comment]
Hello @cmaumet,
If you don't like the compact output of spm_jsonwrite, you can change this behaviour by modifying the code of the function:
`S = jsonwrite_var(json,0); % 0 or NaN`
I will make the option accessible from the input arguments of the function.
I have also modified the JSON export so that you can now use [containers.Map](https://www.mathworks.com/help/matlab/map-containers.html) instead of structures (the downside is that they keys are reordered alphabetically). This is not done for the import.

**Comment by cmaumet** [comment]
Hello @gllmflndn! Having an option to specify the indentation would be great! 

What do you think of the issue of `@` being converted to `x_`? Is this also something that could be handled at the level of `spm_jsonwrite`?

> I have also modified the JSON export so that you can now use containers.Map instead of structures (the downside is that they keys are reordered alphabetically). This is not done for the import.

Am I correct in assuming that this mean that a round-trip comparison using `verifyEqual` on a json document and a copy of this document exported using `spm_jsonwrite` would return True (as opposed to False now)? Or is this a separate issue? 

Thank you!

**Pull request merged**